contracts/src/internal/HyperdriveLP.sol

```diff
@@ -287,21 +329,47 @@ abstract contract HyperdriveLP is
         // excess idle calculation fails, we revert to avoid allowing the system
         // to enter an unhealthy state. A failure indicates that the present
         // value can't be calculated.
-        bool success = _distributeExcessIdleSafe(vaultSharePrice);
+        success = _distributeExcessIdleSafe(vaultSharePrice);
         if (!success) {
             revert IHyperdrive.DistributeExcessIdleFailed();
         }
 
+        // Check to see whether or not adding this liquidity will result in
+        // worsened price discovery. If the spot price can't be brought to one
+        // and price discovery worsened after adding liquidity, we revert to
+        // avoid dangerous pool states.
+        uint256 latestCheckpoint_ = latestCheckpoint; // avoid stack-too-deep
+        uint256 lpShares_ = lpShares; // avoid stack-too-deep
+        IHyperdrive.Options calldata options = _options; // avoid stack-too-deep
+        uint256 vaultSharePrice_ = vaultSharePrice; // avoid stack-too-deep
+        int256 solvencyAfterMaxLongAfter;
+        (
+            solvencyAfterMaxLongAfter,
+            success
+        ) = _calculateSolvencyAfterMaxLongSafe(
+            _marketState.shareReserves,
+            _marketState.shareAdjustment,
+            _marketState.bondReserves,
+            vaultSharePrice_,
+            _marketState.longExposure,
+            _nonNettedLongs(latestCheckpoint_ + _positionDuration)
+        );
+        if (
+            !success ||
+            solvencyAfterMaxLongAfter < solvencyAfterMaxLongBefore.min(0)
```

Comment:
why do we need the min(0)?

Reply:
Good question. I can add a comment to clarify. This is the logic that I added to handle the case where adding liquidity improves price discovery. If solvencyAfterMaxLongBefore is less than 0, we make sure that solvencyAfterMaxLongAfter is larger than solvencyAfterMaxLongBefore. Otherwise, we just make sure it’s greater than zero

Reply:
okay okay i follow it now. it was late and i was reading it wrong. ya this fails if the solvency gets worse (when before and after are negative) or if just after is negative.